### PR TITLE
ci: add batching to cypress e2e tests

### PR DESCRIPTION
**Pull request by maiyerlee**
## What changed

- Organize e2e tests into its own job section for clarity
- Implement running of e2e tests in parallel using a [matrix strategy  ](https://docs.github.com/en/actions/using-jobs/using-a-matrix-for-your-jobs)

The GHA checks now complete in approximately 8 minutes (a significant improvement from over 20 minutes).

## Issue

[2368](https://github.com/HHS/OPRE-OPS/issues/2368)

## How to test

Create draft PRs and examine checks.

## Definition of Done Checklist
- [x] OESA: Code refactored for clarity
- [x] OESA: Dependency rules followed
<del> - [ ] Automated unit tests updated and passed <del>
<del> - [ ] Automated integration tests updated and passed <del>
<del> - [ ] Automated quality tests updated and passed  <del>
<del> - [ ] Automated load tests updated and passed  <del>
<del> - [ ] Automated a11y tests updated and passed  <del>
- [x] Automated security tests updated and passed
<del> - [ ] 90%+ Code coverage achieved  <del>
<del> - [ ] Form validations updated <del>

### CODE REVIEW

File: .github/workflows/e2e_test_reusable.yml [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+        files: ${{ steps.set-files.outputs.FILES }}
+    steps:
+      # Checkout the code
+      - name: Checkout
+        uses: actions/checkout@b4ffde65f46336ab88eb53be808477a3936bae11 #v4.1.1
+
+      # Set the files to run
+      - id: set-files
+        run: |
+          FILE_LIST=$(ls frontend/cypress/e2e/*.cy.js | jq -R -s -c 'split("\n")[:-1]')
```
Comment (maiyerlee) [inline .github/workflows/e2e_test_reusable.yml:19]: | Part | Description |
|------|-------------|
| `ls frontend/cypress/e2e/*.cy.js` | This part lists all files with the `.cy.js` extension in the `frontend/cypress/e2e/` directory. |
| `jq -R -s -c 'split("\n")[:-1]'` | This part of the command processes the output of the `ls` command using `jq`. It reads raw input (`-R`), reads the entire input stream into memory as a single string (`-s`), and outputs the result as a single-line JSON object (`-c`). The `'split("\n")[:-1]'` part splits the input string by newline characters and selects all elements except the last one. This is used to remove the trailing empty string that `jq` adds when splitting. |

File: .github/workflows/e2e_test_reusable.yml [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+        id: get-filename
+        run: |
+          TEST_FILE=$(basename "${{ matrix.file }}" | sed 's/\.cy\.js//')
+          echo "TEST_FILE=$TEST_FILE" >> $GITHUB_OUTPUT
+
+      # Run the Cypress E2E Tests
+      - name: E2E Test
+        working-directory: frontend
+        run: |
+          bun run cypress install && bun run test:e2e --spec $(echo "${{ matrix.file }}" | sed 's/frontend\///')
```
Comment (maiyerlee) [inline .github/workflows/e2e_test_reusable.yml:67]: The `sed` command changes the file path by taking out "frontend/" from the file name

File: .github/workflows/e2e_test_reusable.yml [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+    steps:
+      # Checkout the code
+      - name: Checkout
+        uses: actions/checkout@b4ffde65f46336ab88eb53be808477a3936bae11 #v4.1.1
+
+      # Set the files to run
+      - id: set-files
+        run: |
+          FILE_LIST=$(ls frontend/cypress/e2e/*.cy.js | jq -R -s -c 'split("\n")[:-1]')
+          echo "FILES=$FILE_LIST" >> $GITHUB_OUTPUT
```
Comment (maiyerlee) [inline .github/workflows/e2e_test_reusable.yml:20]: Output looks like:

`["frontend/cypress/e2e/agreementDelete.cy.js","frontend/cypress/e2e/agreementDetails.cy.js",...,"frontend/cypress/e2e/reviewAgreement.cy.js"]`

**Review by johndeange — APPROVED** [review]

**Comment by maiyerlee** [comment]
Per today's Engineers OH discussion, we use checks to ensure PRs are generally ready for the main branch -> we need to know when the first check failure occurs -> thus, the `fail-fast` flag is now set to true in the e2e job

**Pull request merged**